plugins/inputs/hugepages/hugepages.go

```diff
@@ -0,0 +1,291 @@
+		"free_hugepages",
+		"nr_hugepages",
+		"surplus_hugepages",
+	}
+
+	hugepagesMetricsFromMeminfo = []string{
+		"HugePages_Total",
+		"HugePages_Free",
+		"HugePages_Rsvd",
+		"HugePages_Surp",
+		"Hugepagesize",
+		"Hugetlb",
+		"AnonHugePages",
+		"ShmemHugePages",
+		"FileHugePages",
+	}
+)
+
+var hugepagesSampleConfig = `
+  ## Supported huge page types:
+  ##   - "root" - based on root huge page control directory: /sys/kernel/mm/hugepages
+  ##   - "per_node" - based on per NUMA node directories: /sys/devices/system/node/node[0-9]*/hugepages
+  ##   - "meminfo" - based on /proc/meminfo file
+  # hugepages_types = ["root", "per_node"]
+`
+
+type Hugepages struct {
+	HugepagesTypes []string `toml:"hugepages_types"`
+
+	gatherRoot    bool
+	gatherPerNode bool
+	gatherMeminfo bool
+
+	rootHugepagePath string
+	numaNodePath     string
+	meminfoPath      string
+}
+
+func (h *Hugepages) Description() string {
+	return "Gathers huge pages measurements."
+}
+
+func (h *Hugepages) SampleConfig() string {
+	return hugepagesSampleConfig
+}
+
+func (h *Hugepages) Init() error {
+	err := h.parseHugepagesConfig()
+	if err != nil {
+		return err
+	}
+
+	h.rootHugepagePath = rootHugepagePath
+	h.numaNodePath = numaNodePath
+	h.meminfoPath = meminfoPath
+
+	return nil
+}
+
+func (h *Hugepages) Gather(acc telegraf.Accumulator) error {
+	var err error
+
+	if h.gatherRoot {
+		err = h.gatherRootStats(acc)
+		if err != nil {
+			return err
+		}
+	}
+
+	if h.gatherPerNode {
+		err = h.gatherStatsPerNode(acc)
+		if err != nil {
+			return err
+		}
+	}
+
+	if h.gatherMeminfo {
+		err = h.gatherStatsFromMeminfo(acc)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// gatherStatsPerNode collects root hugepages statistics
+func (h *Hugepages) gatherRootStats(acc telegraf.Accumulator) error {
+	return h.gatherFromHugepagePath(h.rootHugepagePath, hugepagesMetricsRoot, "hugepages_"+rootHugepages, nil, acc)
+}
+
+// gatherStatsPerNode collects hugepages statistics per NUMA node
+func (h *Hugepages) gatherStatsPerNode(acc telegraf.Accumulator) error {
+	nodeDirs, err := ioutil.ReadDir(h.numaNodePath)
+	if err != nil {
+		return err
+	}
+
+	// read metrics from: node*/hugepages/hugepages-*/*
+	for _, nodeDir := range nodeDirs {
+		if !nodeDir.IsDir() || !strings.HasPrefix(nodeDir.Name(), "node") {
+			continue
+		}
+
+		nodeNumber := strings.TrimPrefix(nodeDir.Name(), "node")
+		_, err := strconv.Atoi(nodeNumber)
+		if err != nil {
+			continue
+		}
+
+		perNodeTags := map[string]string{
+			"node": nodeNumber,
+		}
+		hugepagesPath := filepath.Join(h.numaNodePath, nodeDir.Name(), "hugepages")
+		err = h.gatherFromHugepagePath(hugepagesPath, hugepagesMetricsPerNUMANode, "hugepages_"+perNodeHugepages, perNodeTags, acc)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (h *Hugepages) gatherFromHugepagePath(hugepagesPath string, possibleMetrics []string, measurementName string,
+	tagsToUse map[string]string, acc telegraf.Accumulator) error {
```

Comment:
You might want to shorten the variable names here. Furthermore, the `acc` should always be first IIRC. So how about
```suggestion
func (h *Hugepages) gatherFromHugepagePath(acc telegraf.Accumulator, measurement, path string, fileFilter []string, defaultTags map[string]string) error {
```

Reply:
Changed.